pl_bolts/models/rl/sac_model.py

```diff
@@ -0,0 +1,421 @@
+        Then calculates loss based on the minibatch recieved
+
+        Args:
+            batch: current mini batch of replay data
+            _: batch number, not used
+            optimizer_idx: not used
+        """
+        policy_optim, q1_optim, q2_optim = self.optimizers()
+        policy_loss, q1_loss, q2_loss = self.loss(batch)
+
+        policy_optim.zero_grad()
+        self.manual_backward(policy_loss)
+        policy_optim.step()
+
+        q1_optim.zero_grad()
+        self.manual_backward(q1_loss)
+        q1_optim.step()
+
+        q2_optim.zero_grad()
+        self.manual_backward(q2_loss)
+        q2_optim.step()
+
+        # Soft update of target network
+        if self.global_step % self.sync_rate == 0:
+            self.soft_update_target(self.q1, self.target_q1)
+            self.soft_update_target(self.q2, self.target_q2)
+
+        self.log_dict({
+            "total_reward": self.total_rewards[-1],
+            "avg_reward": self.avg_rewards,
+            "policy_loss": policy_loss,
+            "q1_loss": q1_loss,
+            "q2_loss": q2_loss,
+            "episodes": self.done_episodes,
+            "episode_steps": self.total_episode_steps[-1]
+        })
+
+    def test_step(self, *args, **kwargs) -> Dict[str, Tensor]:
+        """Evaluate the agent for 10 episodes"""
+        test_reward = self.run_n_episodes(self.test_env, 1)
+        avg_reward = sum(test_reward) / len(test_reward)
+        return {"test_reward": avg_reward}
+
+    def test_epoch_end(self, outputs) -> Dict[str, Tensor]:
+        """Log the avg of the test results"""
+        rewards = [x["test_reward"] for x in outputs]
+        avg_reward = sum(rewards) / len(rewards)
+        self.log("avg_test_reward", avg_reward)
+        return {"avg_test_reward": avg_reward}
+
+    def _dataloader(self) -> DataLoader:
+        """Initialize the Replay Buffer dataset used for retrieving experiences"""
+        self.buffer = MultiStepBuffer(self.replay_size, self.n_steps)
+        self.populate(self.warm_start_size)
+
+        self.dataset = ExperienceSourceDataset(self.train_batch)
+        return DataLoader(dataset=self.dataset, batch_size=self.batch_size)
+
+    def train_dataloader(self) -> DataLoader:
+        """Get train loader"""
+        return self._dataloader()
+
+    def test_dataloader(self) -> DataLoader:
+        """Get test loader"""
+        return self._dataloader()
+
+    def configure_optimizers(self) -> Tuple[Optimizer]:
+        """ Initialize Adam optimizer"""
+        policy_optim = optim.Adam(self.policy.parameters(), self.hparams.policy_learning_rate)
+        q1_optim = optim.Adam(self.q1.parameters(), self.hparams.q_learning_rate)
+        q2_optim = optim.Adam(self.q2.parameters(), self.hparams.q_learning_rate)
+        return policy_optim, q1_optim, q2_optim
+
+    @staticmethod
+    def add_model_specific_args(arg_parser: argparse.ArgumentParser, ) -> argparse.ArgumentParser:
```

Comment:
Consider using `LightningCLI` see: https://devblog.pytorchlightning.ai/auto-structuring-deep-learning-projects-with-the-lightning-cli-9f40f1ef8b36

Reply:
(Same as A2C PR) Hi, I looked a bit into it, and It seems like in order to add the arguments (add_model_specific_args()) in the lightning module, I'll have to write my own subclass of LightningCLI, which defeats the purpose to use it to simplify the code. Let me know if my understanding is wrong. Or is there an example of using it in other RL algorithms?

Reply:
I think it's fine for now!